### PR TITLE
TVI/IOL extractor

### PR DESCRIPTION
**Pull request by tioocnt**
Hi.

Maybe you can consider to add this extractor.

It supports a major Portuguese TV channel (both livestreams and static videos).

The tests might need some work because I can't get the md5 to work only on the file start and because the content seems to change between invocations.

I ran flake8 and got no output.

Feel free to ask for more information.

### CODE REVIEW

**Comment by remitamine** [comment]
you can also extract another m3u8 manifest and mp4 video url:
```python
"http://video-on-demand.iol.pt/vod_http/mp4:" + video_id + "-L-500k.mp4/playlist.m3u8"
"http://www.iol.pt/videos-file/" + video_id + "-L-500k.mp4"
```

**Comment by tioocnt** [comment]
Can't find those sources.

Which urls did you use?

The closet thing I found was using

http://www.tvi24.iol.pt/videos/passos-criacao-de-emprego-e-facto-muito-importante/55f816640cf2e6961770ef7a/2

that has

<meta property="og:video" content="http://www.iol.pt/flashplayers/jwplayer_5_9.swf?file=mp4%3A55f816640cf2e6961770ef7a-L-500k&streamer=rtmp%3A%2F%2Fvideo1.iol.pt%2Fvod&config=http%3A%2F%2Fwww.iol.pt%2Fconfig_fb.xml%3Fsite%3Dtvi24%26scheme%3Dhttp"/>
            <meta property="og:video:secure_url" content="https://www.iol.pt/flashplayers/jwplayer_5_9.swf?file=mp4%3A55f816640cf2e6961770ef7a-L-500k&streamer=rtmp%3A%2F%2Fvideo1.iol.pt%2Fvod&config=https%3A%2F%2Fwww.iol.pt%2Fconfig_fb.xml%3Fsite%3Dtvi24%26scheme%3Dhttps"/>


I should add that this "portal" has a bunch of different layouts and what sometimes is present is present in one is absent in another.

**Comment by remitamine** [comment]
i find them in the js code.
there is also a method to extract rtmp info in the js code:
```javascript
file:       'mp4:'+video_id+'-L-500k',
streamer:   'rtmp://'+server+'/vod',
```
server can be extracted from http://iol.pt/videolb if this fail,than fallback to video1.iol.pt

**Comment by tioocnt** [comment]
OK. the video_id used in the JavaScript is not the actual video_id form the URL. It can be extracted from the m3u8 URL I was already parsing. (I think it corresponds to Multimedia_ID!=video_id)

I've fixed that.

Tests now work and should not break.

I haven't found any rtmp related code. Maybe it's URL dependent.
To fix that I'll need more input.

There's still the possibility of implementing the playlist functionality for series.
I can't do it right now. Maybe for another pull request.

Comments welcomed.

**Comment by remitamine** [comment]
> I haven't found any rtmp related code. Maybe it's URL dependent.
To fix that I'll need more input.

the code for rtmp in this https://github.com/rg3/youtube-dl/pull/7080#issuecomment-146031907 can be found here: http://cdn.iol.pt/js/iol.js at line 75,76.

File: youtube_dl/extractor/iol.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            multimedia_id = match.group(1)
+            m3u8_url_default = 'http://video-on-demand.iol.pt/vod_http/mp4:' + multimedia_id + '-L-500k.mp4/playlist.m3u8'
+            formats_m3u8_default = self._extract_m3u8_formats(m3u8_url_default, video_id, ext='mp4')
+            formats.extend(formats_m3u8_default)
+            formats.append({
+                'url': 'http://www.iol.pt/videos-file/' + multimedia_id + '-L-500k.mp4',
+                'format_id': 'http_500',
+                'tbr': 500,
+                'protocol': 'http',
+                'protocol': 'http',
```
Comment (remitamine) [inline youtube_dl/extractor/iol.py]: duplicate line

File: youtube_dl/extractor/iol.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        title = re.sub(r' \| TVI Player$', '', title, re.IGNORECASE)
+
+        description = self._og_search_description(webpage)
+        thumbnail = self._og_search_thumbnail(webpage)
+        m3u8_url = self._html_search_regex(r'''videoUrl:\s*'([^']+\.m3u8[^']*)'\s*,''', webpage, 'm3u8 playlist')
+        formats = self._extract_m3u8_formats(m3u8_url, video_id, ext='mp4')
+
+        match = re.search(r'smil:([0-9a-f]{24})-L', m3u8_url, re.IGNORECASE)
+        if match:
+            multimedia_id = match.group(1)
```
Comment (remitamine) [inline youtube_dl/extractor/iol.py]: can be extracted directly from the meta tag with name `iol:id` like that:
```html
<meta name="iol:id" content="55f878f90cf203f8b03cea6d" />
```

File: youtube_dl/extractor/iol.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+import re
+
+
+class IOLIE(InfoExtractor):
+    _VALID_URL = r'http://(tviplayer|(www\.tvi24))\.iol\.pt/.*/(?P<id>[0-9a-f]{24})($|\/)'
```
Comment (remitamine) [inline youtube_dl/extractor/iol.py]: maisfutebol.iol.pt is simular to tvplayer.iol.pt and tvi24.iol.pt.
to detect only urls with videos,may be it's better to check that `/video/` or `/videos/` in the url.

File: youtube_dl/extractor/iol.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            'description': u'PM sublinha que é o nível mais elevado de há vários anos'
+        }
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._og_search_title(webpage)
+        title = re.sub(r' \| TVI Player$', '', title, re.IGNORECASE)
```
Comment (remitamine) [inline youtube_dl/extractor/iol.py]: doesn't always contain `| TVI Player`.
can be extracted from the ioplayer title.
for example:
```javascript
.iolplayer({
					autostart: true,
                    videoUrl: 'http://vod-auth.iol.pt/vod/_definst_/smil:55f730c40cf23fa665481b18-L/playlist.m3u8?wmsAuthSign=c2VydmVyX3RpbWU9MTAvMTAvMjAxNSA4OjI5OjIzIEFNJmhhc2hfdmFsdWU9Y1FlYVJYWEovMWR6Z1pUNTllOWJidz09JnZhbGlkbWludXRlcz0xNDQwJmlkPTBmYzE2NGVmLTQ3OTktNDY3Mi1iZTI2LTcyN2FmMzM4NGRhOA==',
                    imageId: '55f737330cf23fa665481b2d',
                    title: 'Isso é tudo muito bonito, mas: Concatena, filho, concatena',
```

File: youtube_dl/extractor/iol.py (hunk 3 above)
Comment (tioocnt) [inline youtube_dl/extractor/iol.py]: There should be a lot of sites like that and not easy to enumerate them all.

Maybe just accept every *.iol.pt site?

Worst case scenario there's no video to extract...

**Comment by tioocnt** [comment]
RTMP isn't always available

I check if cdn.iol.pt/js/iol.js is included and if it is I add rtmp as a valid format.

RTMP download is very unstable, most times rtmpdump either loops forever or just exits with error

maisfutebol.iol.pt added with tests
thumbnail tests working
use iol:id meta with the old method as fallback
now I try to actually parse .iolplayer as json using transform_source=js_to_json but this often breaks due to js code so I fallback to regexps

Feedback welcome

File: docs/supportedsites.md [hunk 5]
```diff
@@ -226,6 +226,7 @@
  - **Instagram**
  - **instagram:user**: Instagram user profile
  - **InternetVideoArchive**
+ - **IOL**: iol.pt sites (tvi archive/stream and maisfutebol)
```
Comment (jaimeMF) [inline docs/supportedsites.md]: This file is autogenerated, you can set the `IE_DESC` property in the extractor instead.

Comment (tioocnt) [inline docs/supportedsites.md]: Fixed.

Thanks for the comment.

I usually don't think files in the repo are auto-generated so I took the liberty.

**Comment by tioocnt** [comment]
Any chance of this being merged?

**Comment by tioocnt** [comment]
Sorry, closed by mistake.

**Comment by yan12125** [comment]
Since #8497 landed, please move changes in ```youtube_dl/extractors/__init__.py``` to ```youtube_dl/extractors/extractors.py```. Check step 5 of the new [developer instructions](https://github.com/rg3/youtube-dl/blob/master/README.md#adding-support-for-a-new-site) for more information.

**Comment by tioocnt** [comment]
Any chance of this ever being integrated?

It still works and I use it regularly.

Not saying there isn't room for improvement but this is a major Portuguese TV station and not being able to download the videos is quite annoying

**Comment by mpenkov** [comment]
@tioocnt Just checking up on this PR. It looks like you have done a lot of work, but the Travis CI tests failed for your changes. Could you have a look and try to make them pass?

Once the tests pass, it'll be a lot easier to convince the maintainers to merge this PR.

**Pull request closed without merging**